### ISSUE TITLE
Improve WFFC handling by removing the dummy pod step

### ISSUE THREAD

**Issue by rmohr** [kubevirt/kubevirt#6531]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>

/kind bug

> /kind enhancement

**What happened**:

The WFFC handling of PVCs right now involves first starting a pod without a VM inside which binds the WFFC PVCs to a node.
Then we hope that when we stop that pod and start the actual pod with a VM inside, that the pod can still fit on the node.

In case that a cluster is used well, this assumption is very likely pretty often false. Also our startup flow is pretty complicated due to that.

Apart from memory and cpu also additional limited resources like devices or hugepages influence how likely it is that a pod can go to the same node again.

**What you expected to happen**:

Instead of using a WFFC pod we should do the following:

 1) record on the VMI status that a DataVolume is used and if it is done with preparataions
 2) record on PVCs if annotations on it indicate that it is a DataVolume
 3) use this status info on the VMI status in virt-handler and virt-launcher to delay timeouts until the import opertions are done
 4) continue with the normal startup flow.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- KubeVirt version (use `virtctl version`):
- Kubernetes version (use `kubectl version`):
- VM or VMI specifications:
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

**Comment by rmohr** [comment]
@awels @mhenriks FYI

**Comment by awels** [comment]
So let me see if I understand this correctly. Instead of the dummy pod, you want to just start the virt-launcher pod normally, but not start the VM process until all the volumes are marked as ready? Starting the virt-launcher pod will bind the PVCs to the node the pod is running on. But instead of starting the VM process, we just wait until all the volumes are marked as ready, which will be ongoing now that the PVCs are bound. So for a small period of time, there will be 2 pods that use the PVCs. The virt-launcher that doesn't do anything because not all volumes are ready, and then the CDI pod that is populating the volume.

For RWO in theory this should work, as according to the documentation RWO is on a node level, so multiple pods on the same node should be able to access the volume. However some provisioners might not allow multiple pods even if they are on the same node. I am trying to remember which one gave us trouble with that in the past @mhenriks maybe you remember?

**Comment by rmohr** [comment]
Yes that would be the rough idea. I was counting on the RWO behaviour which you mention. If we can't assume that storage providers adhere to that, then we probably can't do that (still wonder if we can live with e.g. 5% of exotic providers not working).

Apart from the scheduling issues, I had troubles implementing some advanced scheduling features due to this (around NUMA). I eventually could solve it, but our architecture could profit a lot if we could get rid of this dummy pod. I would therefore love to see that we really consider an alternative and investigate if we can go another route.

**Comment by alicefr** [comment]
Could we deploy the CDI importer container inside the virt-launcher pod as init container?

**Comment by rmohr** [comment]
> WFFC pod not necessary with Volume Populators (currently alpha).
> 
> But Volume Populators will only work with CSI provisioners. So provisioners like local storage and hpp (classic) are still potentially problematic.

So sounds like at some point we can drop the complicated flow because the CSI implementations can make use of the populator and for the classic types we can rely on the node-level-locking behaviour?

**Comment by awels** [comment]
Yes, I think that would be the appropriate approach.

**Comment by npu21** [comment]
@rmohr , Could you please take a look at this issue? The pv created on node hci-002 which should created on hci-001.
https://github.com/kubevirt/containerized-data-importer/issues/2192

**Comment by xuzhenglun** [comment]
`ReadWriteOncePod` is beta in 1.26.  The discussion about this feature has clarified the distinction between RWO (Read-Write Once) and RWOP (Read-Write One Pod), specifically that RWO can be read and written by multiple Pods on a Node, while RWOP can only be read and written by a single Pod on a Node.  Maybe we can proceed with this idea based on this clarification?

**Comment by mhenriks** [comment]
@xuzhenglun I'm not sure how `ReadWriteOncePod` applies here.  It is orthogonal to `WaitForFirstConsumer`.

**Comment by mhenriks** [comment]
KubeVirt now supports CDI with populators.

https://github.com/kubevirt/kubevirt/pull/9812
https://github.com/kubevirt/kubevirt/pull/9918
https://github.com/kubevirt/kubevirt/pull/10065

**Comment by xuzhenglun** [comment]
> I'm not sure how ReadWriteOncePod applies here. It is orthogonal to WaitForFirstConsumer.

I would have thought that with the explicit semantics of RWOP and RWO, we could proceed with the scenario described above:

  > Starting the virt-launcher pod will bind the PVCs to the node the pod is running on. But instead of starting the VM process, we just wait until all the volumes are marked as ready, which will be ongoing now that the PVCs are bound. So for a small period of time, there will be 2 pods that use the PVCs. 

---
However, due to Volume Populator have been supported, this proposal seems unnecessary to continue. Cause Volume Populator is indeed a better solution.
